v1/src/main/java/com/google/cloud/teleport/spanner/AvroRecordConverter.java

```diff
@@ -374,6 +384,45 @@ static Optional<List<String>> readStringArray(
     }
   }
 
+  @VisibleForTesting
+  @SuppressWarnings("unchecked")
+  static Optional<List<Float>> readFloat32Array(
+      GenericRecord record, Schema.Type avroType, String fieldName) {
+    Object fieldValue = record.get(fieldName);
+    if (fieldValue == null) {
+      return Optional.empty();
+    }
+    switch (avroType) {
+        // For type check at compile time, the type of x has to be specified (as cast) so that
+        // convertability to float can be verified.
+      case FLOAT:
+        return Optional.of((List<Float>) fieldValue);
+      case INT:
+        // Can cause loss of precision: un-representable integers will be rounded to the nearest
+        // float.
+        {
+          List<Integer> value = (List<Integer>) fieldValue;
+          return Optional.of(
+              value.stream()
+                  .map(x -> x == null ? null : Float.valueOf(x))
+                  .collect(Collectors.toList()));
+        }
```

Comment:
When will this case occur?  When a customer wants to read a Avro integer into a Float32 in Spanner? 

When you say `un-representable integers` do you mean bigint type of numbers which lie outside of Float32's range?

Reply:
> When will this case occur? When a customer wants to read a Avro integer into a Float32 in Spanner?

Yes, that's correct.

>When you say un-representable integers do you mean bigint type of numbers which lie outside of Float32's range?

No, I'm referring to the int32 type here (IIUC avro integer is an int32)

Since not all int32 values can be represented as a float32 ([wikipedia has some details](https://en.wikipedia.org/wiki/Single-precision_floating-point_format#Precision_limitations_on_integer_values)), rounding is inevitable during the conversion from int32 to float32.

Similar is the case when int64 values are attempted to be parsed as a float64 value -- some int64 values will be rounded off. Ideally that conversion should also have a warning.

Reply:
Thanks for the details.
So in the `DatastreamToSpanner` details, we are taking the approach of throwing exceptions on hidden lossy castings. These templates are often used for migrations, and the thought process is to force the user to make an explicit decision about such lossy casting. 

A recent example of this is #1386 where we will error out the record on lossy conversions. 


Guess I'm trying to understand - 
What is a practical usecase a customer would want to map Avro int32 bit into a float32 while being okay with rounding?

@darshan-sj you have more context than me on Import/Export, do you have any suggestions? What approach we have followed till now for other data types with similar rounding constraints?

Reply:
Thanks, I've removed the INT -> FLOAT conversion for float32. I think this will have to be handled for FLOAT64 too. Not doing it in this PR.

Reply:
Import/Export is majorly used as Backup/restore tool and we don't support transformations. So, in majority of the cases from and to datatypes would be same since we create the schema in Spanner according to the datatype of the avro record. Only case where we would reach here would be if customer created the tables in spanner manually and changed datatype of a column from INT -> FLOAT and then run import on that. I think it's okay to have lossy conversion since customer is explicitly changing the datatype.